File: wrappers/javascript/aries-askar-shared/src/store/Scan.ts
import type { Entry, EntryObject } from './Entry'
import type { Store } from './Store'
import type { EntryListHandle, ScanHandle } from '../crypto'

import { ariesAskar } from '../ariesAskar'
import { AriesAskarError } from '../error'

import { EntryList } from './EntryList'

export class Scan {
  private _handle?: ScanHandle
  private _listHandle?: EntryListHandle
  private store: Store
  private profile?: string
  private category: string
  private tagFilter?: Record<string, unknown>
  private offset?: number
  private limit?: number

  public constructor({
    category,
    limit,
    offset,
    profile,
    tagFilter,
    store,
  }: {
    profile?: string
    category: string
    tagFilter?: Record<string, unknown>
    offset?: number
    limit?: number
    store: Store
  }) {
    this.category = category
    this.profile = profile
    this.tagFilter = tagFilter
    this.offset = offset
    this.limit = limit
    this.store = store
  }

  public get handle() {
    return this._handle
  }

  private async forEach(cb: (row: Entry, index?: number) => void) {
    if (!this.handle) {
      if (!this.store?.handle) throw AriesAskarError.customError({ message: 'Cannot scan from closed store' })
      this._handle = await ariesAskar.scanStart({
        storeHandle: this.store.handle,
        limit: this.limit,
        offset: this.offset,
        tagFilter: this.tagFilter,
        profile: this.profile,
        category: this.category,
      })
    }

    try {
      let recordCount = 0
      // Loop while limit not reached (or no limit specified)
      while (!this.limit || recordCount < this.limit) {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const listHandle = await ariesAskar.scanNext({ scanHandle: this._handle! })
        if (!listHandle) break

        this._listHandle = listHandle
        const list = new EntryList({ handle: this._listHandle })

        recordCount = recordCount + list.length
        for (let index = 0; index < list.length; index++) {
          const entry = list.getEntryByIndex(index)
          cb(entry)
        }
      }
    } finally {
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      ariesAskar.scanFree({ scanHandle: this._handle! })
    }
  }

  public async fetchAll() {
    const rows: Array<EntryObject> = []
    await this.forEach((row) => rows.push(row.toJson()))
    return rows
  }
}

File: wrappers/kotlin/src/commonMain/kotlin/Store/Scan.kt
package askar.Store

import aries_askar.askar_get_current_error
import askar.Askar
import askar.crypto.EntryListHandle
import askar.crypto.ScanHandle
import kotlinx.cinterop.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject

@OptIn(UnsafeNumber::class)
class Scan(
    private val profile: String? = null,
    private val category: String,
    private val tagFilter: JsonObject = buildJsonObject {},
    private val offset: Int = 0,
    private val limit: Int = -1,
    private val store: Store
) {
    private var handle: ScanHandle? = null
    private var listHandle: EntryListHandle? = null

    fun handle(): ScanHandle? {
        return this.handle
    }

    suspend fun forEach(cb: (row: Entry, index: Int) -> Unit) {
        memScoped {
            if (handle == null) {
                handle = Askar.scan.scanStart(store.handle().handle, limit, offset, tagFilter, profile, category)
            }
            try {
                var recordCount = 0
                while (limit == -1 || recordCount < limit) {
                    val list = Askar.scan.scanNext(handle!!.handle, this) ?: break
                    listHandle = list

                    val entryList = EntryList(list)

                    recordCount += entryList.length()
                    for (i in 0 until entryList.length()) {
                        val entry = entryList.getEntryByIndex(i)
                        cb(entry, i)
                    }
                }

            } finally {
                Askar.scan.free(handle!!.handle)
            }
        }

    }

    suspend fun fetchAll(): ArrayList<EntryObject> {
        val rows = ArrayList<EntryObject>()
        forEach { row: Entry, _ -> rows.add(row.toJson()) }
        return rows
    }
}